README.md
# GFR

## How to run tests
The following script will run the tests and generate reports using ```coverage.py```. The reports are ```xml``` files and can be interpreted with e.g. the VS Code extension; Coverage Gutter
```
(venv)> ./run-tests.sh
```
Remember to make the script executable using:
```
(venv)> chmod a+x ./run-tests.sh
```

---
## Install, Deploy, Run
### General dependencies
The project was built using:
* python 3.6
* pip3
* dcmtk v3.6.2 (https://dicom.offis.de/dcmtk.php.en)
  * findscu
  * getscu
  * storescu 
* Nginx (+ uWSGI)

The project was built and tested on CentOS 7 and Ubuntu 18.04 (bionic beaver)

### Installing and using virtualenv
Virtualenv is install through pip3:
```
> pip3 install virtualenv
```

Initializing virtualenv
```
> virtualenv venv
```

Running virtualenv:
```
> source venv/bin/activate
```

Exitting the virtualenv:
```
> deactivate
```

### Installing requirements
Once in the virtualenv, the required python packages can be installed by running:
```
(venv)> pip install -r requirements.txt
```

#### Installing new packages
Whenever a new package is added the ```requirements.txt``` must be updated. This can be done by running:
```
(venv)> pip freeze > requirements.txt
```
and then adding it to git (with a ```git add``` and ```git commit```).

### Running in debug mode
The Django comes with a development/debugging minimal web server, which can be ran by setting the variable ```debug=True```, under the ```settings.py``` file in the main project directory.

Once this is done, the web development web server can be ran using:
```
(venv)> python manage.py runserver
```

#### Debug mode on local network
To allow hosts on the local network to access the debug test site, run the command:
```
(venv)> python manage.py runserver 0.0.0.0:8000
```

### Deploying with nginx
1. Install uwsgi:
```
> pip3 install uwsgi
```
2. Test that the Django application can run using uwsgi
```
(venv)> uwsgi --http :8000 --module clairvoyance.wsgi
```
3. Install nginx (**Ubuntu**)
```
> sudo apt install nginx
```
3. Install nginx (**CentOS 7**)
```
> sudo yum install epel-release
> sudo yum install nginx
```
4. Test that nginx was correctly installed
```
sudo systemctl start nginx
```
Then goto localhost:80 in a browser, this should display a welcome message.

5. Configure nginx
Download the file: https://github.com/nginx/nginx/blob/master/conf/uwsgi_params, and place it in the main Django project directory.

If the /etc/nginx/sites-available and /etc/nginx/sites-enabled directories doesn't exist, then create them and edit the file /etc/nginx/nginx.conf and add the line:
```
include /etc/nginx/sites-enabled/*;
```
(No need to include the sites-avaiable since sites-enabled will contain symlinks to configs inside this directory)

Write the following to a config file under /etc/nginx/sites-available/clairvoyance_nginx.conf
```
# clairvoyance_nginx.conf

# the upstream component nginx needs to connect to
upstream django {
    server unix:///home/simon/Documents/clearance-stuff/GFR/clairvoyance.sock; # for a file socket
}

# configuration of the server
server {
    listen      80; # the port your site will be served on
    server_name 193.3.238.103; # substitute your machine's IP address
    charset     utf-8;

    # max upload size
    client_max_body_size 75M;

    location /static {
        alias /home/simon/Documents/clearance-stuff/GFR/main_page/static; # your Django project's static files
    }

    # Finally, send all non-media requests to the Django server.
    location / {
        uwsgi_pass  django;
        include     /home/simon/Documents/clearance-stuff/GFR/uwsgi_params; # the uwsgi_params file you installed
    }
}
```
Allow nginx to see this file by creating a symlink to it under the directory: /etc/nginx/sites-enabled/ by using the following command:
```
> sudo ln -s /etc/nginx/sites-available/clairvoyance_nginx.conf /etc/nginx/sites-enabled/clairvoyance_nginx.conf
```

Edit the file /etc/nginx/default, so that nginx runs on port 8081 by default.

6. Restart the nginx service:
```
> sudo /etc/init.d/nginx restart
```
7. Test that the server can run using:
```
> uwsgi --socket clairvoyance.sock --module clairvoyance.wsgi --chmod-socket=666
```
Now you should be able to go to 'localhost' in a browser and see the server running (we can use just 'localhost' since we are running on port 80)

---

## Example of user creation
```
(venv)> python manage.py shell
>>> from main_page.models import User, Config
>>> c = Config(config_id=1, accepted_procedures='Clearance Fler-blodprøve^GFR, Cr-51-EDTA, one sampel^GFR, Tc-99m-DTPA', rigs_aet='VIMCM', rigs_ip='10.143.128.247', rigs_port='3320', rigs_calling='RH_EDTA', pacs_aet='TEST_DCM4CHEE', pacs_ip='127.0.0.1', pacs_port='11112', pacs_calling='RH_EDTA')
>>> u = User(id=1, username='rh_test', hospital='RH', config=c)
>>> u.set_password('rh_test')
>>> c.save()
>>> u.save()
```

The creation can be validated using the sqlitebrowser, where the users lie under the table called "main_page_user" and user configurations under "main_page_config".

---

## Notes
Example of different user agents (```request.META['HTTP_USER_AGENT']```):
Internet explorer on the RH laptop: 
Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko

Firefox on Ubuntu 18.04: 
Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko/20100101 Firefox/66.0

Chrome on Ubuntu 18.04: 
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36


## Code documentation
...to come in the future...

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
<!-- and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). -->

## 1.1 TODO Features
  - [ ] Resolve timeout issues related to retreiving the history of studies (compress the images uploaded to PACS)

---

## 1.0.6 - 2019-10-16
### Added
  

### Changed
  

### Fixed
  - Resolve data overwriting issue with ris_thread and ris_query_wrapper, where they could overwrite previously filled out data

---

## 1.0.5 - 2019-10-11
### Added
  - Javascript checks on entered study date to see if they differ from the injection date

### Changed
  - Procedure filtering is blacklisting instead of whitelisting

### Fixed
  - Exporting study results to csv now exports to .csv filetype.

---

## 1.0.4 - 2019-09-04
### Added
  -  csv export, on site single study

### Changed
  - Spelling of sample

### Fixed
  - Saving of the daily thining factor is initially uncheck if already entered previously.
  - Made SambaBackupEndpoint and StudyEndpoint have login required.

---

## 1.0.3 - 2019-08-26
### Added
  - Sorting of admin panel table.
  - Fetching backup files now support Hidex files.

### Changed
  - Resolve minor spelling mistake in tooltip for delete and edit buttons on admin panel.
  - Disabled enter keypress on all input fields on fill_study
  - Deletion and recovery of studies is now in the RESTful api instead of /ajax.
  - Increased thinning factor threshold to 10k

### Fixed
  - Made deletion and recovery of studies work again

---

## 1.0.2 - 2019-08-23
### Added

### Changed
  - Fixed a few Spelling mistakes
  - QA plot changed from a log plot to a Exponentail plot

### Fixed
  - Auto Character add no longer adds an extra character if the character being added is the same as the function would add.
  - Updated Color plot such that it reflect the underlying method
  - Resolved issue in ```calc_clearance``` under clearance_math.py with days to minutes conversion for tests done over multiple days

---

## 1.0.1 - 2019-08-21
### Added
- 'Tilføj tom prøve' button which makes it possible to add tests without samples from a counter.
- Separator line in result graph
- Warning if there is a difference between the injection date and sample date.

### Changed
- Result graph displays integer values for; 'højde', 'GFR', 'GFR, normaliseret' og 'Nyrefunktion ift. Reference Patient'. 
- ```kidney_function``` under clearance_math.py no longer computes age and gender based on CPR number.
- Name and cpr number fields in fill_study are readonly

### Fixed
- Correction of minor spelling errors
- Made 'Stardardtælletal' field have a corresponding label, instead of the strong tag
- Added support for RIS/Accession numbers in all formats
- Sample dates are now correctly converted to the internal format, to avoid problems with computing the time difference between injection time and sample time.


test-site.py
import requests
from subprocess import check_output
import pydicom


test_user = {
  'csrfmiddlewaretoken': [''], 
  'username': ['rh_test'], 
  'password': ['rh_test'], 
  'hospital': ['RH']
}

rigs_numbers = [
  'REGH13727725',
  'REGH13716821',
]

tests = [
  {
    'age': ['77'],
    'calculate': ['Beregn'],
    'cpr': ['131041-0272'],
    'csrfmiddlewaretoken': [''],
    'height': ['158'],
    'injection_date': ['2019-01-24'],
    'injection_time': ['09:20'],
    'name': ['Frederikke Lerche Andersen'],
    'sex': ['Dame'],
    'std_cnt': ['7525'],
    'study_date': ['2019-01-24', '2019-01-24'],
    'study_time': ['12:45', ''],
    'study_type': ['0'],
    'test_value': ['216'],
    'thin_fac': ['980'],
    'vial_weight_after': ['3.112'],
    'vial_weight_before': ['4.094'],
    'weight': ['55']
  },
  {
    'age': ['71'],
    'calculate': ['Beregn'],
    'cpr': ['	280247-1044'],
    'csrfmiddlewaretoken': [''],
    'height': ['164'],
    'injection_date': ['2018-12-13'],
    'injection_time': ['10:20'],
    'name': ['Lise Birte Dreyer'],
    'sex': ['Dame'],
    'std_cnt': ['11579'],
    'study_date': ['2018-12-13', '2018-12-13', '2018-12-13', '2018-12-13', '2019-01-24'],
    'study_time': ['14:21', '14:42', '15:00', '15:21', ''],
    'study_type': ['2'],
    'test_value': ['183', '168', '161', '155'],
    'thin_fac': ['917'],
    'vial_weight_after': ['3.1070'],
    'vial_weight_before': ['3.7770'],
    'weight': ['73']
  },
]

def get_csrftoken(session, url):
  session.get(url)
  if 'csrftoken' in session.cookies:
    return session.cookies['csrftoken']
  else:
    return session.cookies['csrf']

def main(): 
  # Create login session
  base_url = 'http://localhost:8000'
  
  s = requests.Session()
  
  test_user['csrfmiddlewaretoken'][0] = get_csrftoken(s, base_url)
  s.post(base_url, data=test_user)

  # Send test payloads
  for rigs_nr, payload in zip(rigs_numbers, tests):
    exam_url = '{0}/fill_study/{1}'.format(base_url, rigs_nr)
    s.get(exam_url)

    payload['csrfmiddlewaretoken'][0] = get_csrftoken(s, exam_url)

    s.post(exam_url, data=payload)

  # Send the output dicom objects to the dcm4chee database
  obj_paths = ['./Active_Dicom_objects/RH/{0}.dcm'.format(x) for x in rigs_numbers]
  
  # Set the StudyDate for this specific person
  ds = pydicom.dcmread(obj_paths[1])
  ds.StudyDate = '20181213'
  ds.SOPInstanceUID = '1.3.51.0.1.1.10.143.20.159.13716821.6705280'
  ds.SOPClassUID = '1.2.840.10008.5.1.4.1.1.7'
  ds.save_as(obj_paths[1])

  store_query = [
    'storescu',
    '-aet',
    'RH_EDTA',
    '-aec',
    'TEST_DCM4CHEE',
    '127.0.0.1',
    '11112',
  ] + obj_paths

  check_output(store_query)


if __name__ == '__main__':
  main()